### PR TITLE
Parallelize build system

### PR DESCRIPTION
**Pull request by madig**
This revamps the lower levels of the build system to

1. Simplify targets
2. Cut static instances directly from variable fonts
3. Make make go parallel

Cutting statics from VFs is significant and may be buggy still. It hopefully solves the problem of some features not variating because interpolating UFOs does not interpolate features.fea.

Fontbakery found:
1. `OS/2.fsSelection` is not set properly https://github.com/googlefonts/googlesans/pull/279/checks?check_run_id=2112178808#step:8:166 -- a known issue https://github.com/fonttools/fonttools/issues/1264#issuecomment-422425249
2. `OS/2.xAvgCharWidth` may need to be recomputed? https://github.com/googlefonts/googlesans/pull/279/checks?check_run_id=2112178808#step:8:249
3. Some glyph points are outside their computed bbox? https://github.com/googlefonts/googlesans/pull/279/checks?check_run_id=2112178808#step:8:258

### CODE REVIEW

**Comment by madig** [comment]
Since statics need VFs, maybe we can drop the distinction between static and variable CI?

**Comment by chrissimpkins** [comment]
What is the compile time improvement with the changes here?  The static and VF were compiled and tested in parallel through GH Actions.  I have not looked at the data in detail but my experience was that statics took another ~ 60 sec or so to run through the full CI compile and testing process.

>  It hopefully solves the problem of some features not variating because interpolating UFOs does not interpolate features.fea.

IIUC, benefits are faster compile times to static + VF format artifacts, maybe address some bugs but this is hypothetical at this stage?  Downside is that we are introducing changes in the static build format files.  It isn't completely clear if all of these will be regressions, but some are as you note above. 

This introduces a set of changes relative to production fonts that we will need to investigate.  Static fonts are the default build format in production across most distribution areas.  This is still the case after our GS v4.000 updates.  IIUC the Fonts API will slice static named instances from the VF once the variable format lands in production there.  But it isn't the default yet and there are other distribution areas that will require statics for the foreseeable future.  Statics aren't quite ready to become second class citizens.  We'll need to run through a detailed regression review to make the change.

**Comment by madig** [comment]
On my machine, build time for VF plus statics is reduced by ~60%, from 4 mins down to ~1:20 mins. The change would make features interpolate correctly, a real advantage over the current way, and one that may help us with more complicated SEA scripts that need per-master `pos` statements in their features.fea.

I'll do some TTX dump diffs to assess the changes.

**Comment by madig** [comment]
Here are some findings:

- When removing overlaps, `instancer` also decomposes and removes overlaps in composites (`Aring`, `Ccedilla`, ...). After subsetting, this also means fewer glyphs but more outlines.
- Instances cut from VFs may have more (compatibility) points or are differently ordered contours because there are no "optimizations" done to the VF outlines
- Cut instances are ~20 KB larger than the interpolated ones in the main branch (due to overlap removal?).
- Many components have the USE_MY_METRICS (0x200) flag set
- The STAT table is left in and the name table contains variable axis name leftovers.
- The rvrn feature is left in in opsz=18 fonts, i.e. in fonts where in the VF it is active. In opsz=17 fonts, it's left out.
- Some data needs to be fixed up or inserted afterwards: OS/2.fsSelection, OS/2.xAvgCharWidth, OS/2.panose
- E.g. the Medium style has head.flags bit 1 missing ("Left sidebearing point at x=0"), maybe because...
- ...some advance widths are off by one (Medium, yen.tf)
- GDEF ligature caret positions are indeed different now in e.g. the Medium
- The name tables are generated from the STAT tables, which differ from the custom family names etc. defined in the Designspace
    - E.g. Text Bold: name id3 and id6 say "GoogleSans-TextBold" instead of "GoogleSansText-Bold"
    - E.g. Text Medium: has name id16 and 17 "Google Sans" and "Text Medium" instead of "Google Sans Text" and "Medium"
- The Designspace says axis name "Size" but the STAT table was filled in with "Optical Size"

In a way, instances cut from VFs are really the VFs with the unwanted regions cut out :)

File: requirements.in [hunk 1]
```diff
@@ -1 +1,2 @@
-fontmake
\ No newline at end of file
+fontmake[pathops]
+statmake
```
Comment (chrissimpkins) [inline requirements.in:2]: My understanding is that Dave feels that Google projects should stay with the fontTools STAT table write approach based on recent comments.  What is the rationale for the change to statmake here?

Comment (madig) [reply requirements.in:2]: It's easier and consistent because you write the data once and statmake will do the right thing. I wanted to reduce the lines I needed to write into the Makefile per instance.

File: requirements.txt [hunk 2]
```diff
@@ -2,25 +2,70 @@
...
-fonttools[lxml,ufo,unicode]==4.19.1  # via booleanoperations, cffsubr, compreffor, cu2qu, fontmake, fontmath, glyphslib, ufo2ft, ufolib2
-fs==2.4.11                # via fonttools
-lxml==4.6.2               # via fonttools
-glyphslib==5.3.2         # via fontmake
-pyclipper==1.2.0          # via booleanoperations
-pytz==2020.1              # via fs
-six==1.15.0               # via fs
-ufo2ft[cffsubr]==2.19.2   # via fontmake
-ufolib2==0.8.0            # via fontmake, glyphslib
+appdirs==1.4.4
```
Comment (chrissimpkins) [inline requirements.txt:7]: Hooah.  This requirements.txt file diff gives me chest pain.  We should have a look at the pre/post diff with all of these build requirements changes alone. :)

**Comment by madig** [comment]
FWIW, I decided to leave the PR here for maybe later, as it may be too disruptive right now. Shame about the parallelism though...

**Comment by chrissimpkins** [comment]
> FWIW, I decided to leave the PR here for maybe later, as it may be too disruptive right now. Shame about the parallelism though...

All of the changes here are requirements to transition to the more parallel approach that you have in mind?  Can we look into the build dependency changes, instancing from VF's, and the transition to new tooling for e.g., STAT table writes in separate PR's?  There is a lot going on in this PR.

**Comment by madig** [comment]
All changes are in services of making the Makefile more managable. E.g. the scripts worked by globbing, but should work solely on arguments to be able to work in parallel without interference. I think I can probably keep the current data/interpolation flow and get parallelism by altering it, i.e. making a step that fully loads a DS and pickles it, so it can be picked up by other processes.

**Comment by chrissimpkins** [comment]
> All changes are in services of making the Makefile more managable. E.g. the scripts worked by globbing, but should work solely on arguments to be able to work in parallel without interference. I think I can probably keep the current data/interpolation flow and get parallelism by altering it, i.e. making a step that fully loads a DS and pickles it, so it can be picked up by other processes.

Don't abandon it. I love the idea.  If you want to pursue it, maybe pull out bite sized pieces and open those as smaller scale PR's until we reach the build system of your dreams.  

I suppose my suggestions at this stage of changes here are that you (1) understand how much performance gain there is on the GH Actions boxes where our production build and CI runs (likely limited to 4 CPU and not the Ryzen double digit processor numbers that you have locally if you are creating something that is really, really parallel); (2) understand how much complexity will be added to the build process to achieve those gains.  (1) is to prevent you from wasting your time; (2) is to prevent us from creating a monster that someone else inherits down the road

**Comment by madig** [comment]
Ok, will do. I'm less concerned about CI savings and more about local savings because I run `make` locally and _often_ when testing stuff :)

**Comment by chrissimpkins** [comment]
> Ok, will do. I'm less concerned about CI savings and more about local savings because I run `make` locally and _often_ when testing stuff :)

This means that there is a healthy tension here.  Let's see what we can do to improve your local build workflow and the project simultaneously.  

Google pays for CPU time through the GH Enterprise account so this would reduce financial costs.  We (or maybe just I if you are building locally) sit around and wait for our `build-*` branch artifacts to be pushed on test commits so it addresses that issue.  It will make the remote CI tests faster and we rely on those to determine the ability to merge to main.  There are lots of possible gains.   If these were simple changes that did not affect binaries or maintainability of the source, I'd say it is a no brainer.  We know that it affects the binaries.  I do not understand what the impact will be on build system complexity yet.   Let's keep exploring those tradeoffs and iterate to find something that fits.

**Comment by madig** [comment]
Ok.

Note: using pickling to partition the work to be done would mean we can't use fontmake, because it does not support pickled data as input and probably shouldn't. It is easy to use standalone scripts though, as I have done for Cantarell. That's suboptimal if we want to keep use fontmake as the go-to tool, though.

**Comment by anthrotype** [comment]
👎  on custom scripts, better try contributing any perf improvements upstream

**Comment by chrissimpkins** [comment]
> using pickling to partition the work to be done would mean we can't use fontmake

fontmake is a requirement

**Comment by chrissimpkins** [comment]
OK to close per Niko
